components/logins/android/src/main/java/mozilla/appservices/logins/DatabaseLoginsStorage.kt

```diff
@@ -24,6 +24,27 @@ import mozilla.components.service.glean.private.CounterMetricType
 import mozilla.components.service.glean.private.TimingDistributionMetricType
 import mozilla.components.service.glean.private.LabeledMetricType
 
+/**
+ * This component can emit metrics via Glean, but the consuming application
+ * needs to explicitly opt-in. It does so by calling this function to tell
+ * us what version of our metrics it wants to opt in to.
```

Comment:
I think this makes things much more complicated when analysing the data. See [this Fenix issue](https://github.com/mozilla-mobile/fenix/issues/5626#issuecomment-535985971). From that message:

> For example, when correlating measurements that may or may not have been collected, will we have a way of telling whether something isn't there because it didn't happen or because the user opted out

Is there a specific, concrete use case you're trying to address with this that is currently open? Or is this attempting to resolve the problem in advance?

Metrics versioning adds another layer of confusion to the analysis, which I'd strongly discourage doing. The tables that will contain the ingested data will contain all of these fields, no matter the version.

cc @fbertsch @mdboom

Reply:
For additional clarity, in the current form, the solution also breaks documentation, since versions should be mentioned for metrics.

I would encourage all of us to properly discuss this and make it a Q1 objective.

Reply:
These are good points, thanks!

It's also possible that the way I've named and described this makes it sound more fragile than it is, so just to be clear, here's how I imagined this being used in practice:

* Apps that are already using the logins component (Lockwise, soon to be Fenix depending on whether our metrics land before or after their initial logins integration) can upgrade to newer versions of the logins component without getting any surprise new metrics that may not have been data-reviewed etc.
* Probably with some prodding from our team, those apps do a data-review for the new metrics and update their app to call `logins.enableTelemetry(1)` at startup. As of the next release of the app, they start emitting those metrics.
   * Importantly, they ***do not*** pref individual collection of these metrics on or off at runtime; every version of the app as of this release includes the metrics in their telemetry, just like if it had added some new metrics into its own `metrics.yaml` file.
* If we change the way metrics works in the logins component in the future, we go through the above cycle again, but this time to new version of the app needs to change to call `logins.enableTelemetry(2)` at startup.

Using named constants instead of integers might make it a bit clearer, and maybe "version" is not the best name. But that's the idea, a kind of application-level once-and-done opt-in for "yes we're ready for these metrics to start flowing now" rather than a runtime user-level opt-in-or-out scenario.

> Metrics versioning adds another layer of confusion to the analysis, which I'd strongly
> discourage doing. The tables that will contain the ingested data will contain all of
> these fields, no matter the version.

Does the usage described above change the possibility for confusion here? In my head it doesn't seem any different to adding new metrics directly into a new version of an app, but I'm sure there's plenty of nuance involved that I don't really understand well.

> I would encourage all of us to properly discuss this and make it a Q1 objective.

👍 certainly happy to meet and discuss. I don't have a good sense of your team's schedule, would you be happy to set something up with the right people and timezones @Dexterp37? Ideally we can involve @grigoryk but I also feel like I have a good understanding of his requirements and could represent them if necessary.

Reply:
> every version of the app as of this release includes the metrics in their telemetry,
> just like if it had added some new metrics into its own metrics.yaml file.

I guess one difference here is, you could have a situation where:

* Logins Component vX has landed some metrics, declaring them in its yaml file.
* Lockwise App vY has updated to Logins Component vX, but not enabled the new metrics.
* Transitively scanning the `metrics.yaml` files for Lockwise App vY and its dependencies makes it look like the app *should* be emitting logins-component metrics, but it isn't.

Which I guess in turn, is what leads to the "The tables that will contain the ingested data will contain all of these fields" situation.

Reply:
> 
> 
> These are good points, thanks!

Thank you for your patience on this :) I've also contributed with a more general discussion [here](https://github.com/mozilla-mobile/android-components/pull/5223#issuecomment-562042971).
 
> It's also possible that the way I've named and described this makes it sound more fragile than it is, so just to be clear, here's how I imagined this being used in practice:
> 
>     * Apps that are already using the logins component (Lockwise, soon to be Fenix depending on whether our metrics land before or after their initial logins integration) can upgrade to newer versions of the logins component without getting any surprise new metrics that may not have been data-reviewed etc.

That's a good point. I think that this should be called out in the _changelog_ in Android Components and in here. People updating should be aware and should be consulting the changelog anyway.

>     * Probably with some prodding from our team, those apps do a data-review for the new metrics and update their app to call `logins.enableTelemetry(1)` at startup. As of the next release of the app, they start emitting those metrics.
>       
>       * Importantly, they _**do not**_ pref individual collection of these metrics on or off at runtime; every version of the app as of this release includes the metrics in their telemetry, just like if it had added some new metrics into its own `metrics.yaml` file.

They are _required_ to do the data review for adding the Glean component. Yes, this might need some help by your team to make sure that happens, given that we currently have no build-time way to enforce this. We're trying to address it, but it's a bit complicated. See [this bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1598247).

>     * If we change the way metrics works in the logins component in the future, we go through the above cycle again, but this time to new version of the app needs to change to call `logins.enableTelemetry(2)` at startup.

I *think* that should not be required, unless you merge some concerning metric in your component. And, if you do that, we have much bigger problems :-)

In order to transition this from "I think" to "that's our policy" we reached out to Data Stewardship to hopefully encode these practices.

Other components are currently working this way and it would be *really hard* for them to work differently than that. One prime example could be `engine-gecko-*` components: the [metrics coming from that component](https://github.com/mozilla-mobile/android-components/blob/master/components/browser/engine-gecko-nightly/docs/metrics.md) live in [mozilla-central](https://searchfox.org/mozilla-central/source/toolkit/components/telemetry/geckoview/streaming/metrics.yaml) (see [here](https://www.a2p.it/wordpress/tech-stuff/mozilla/geckoview-glean-fenix-performance-metrics/) for how it works :D ).

Gecko metrics seem to have an high churn rate, with metrics being removed/added quite frequently. I'm not sure products would want to file a data-review on a daily basis, or maybe twice a day, once every new GeckoView nightly comes out.

> Using named constants instead of integers might make it a bit clearer, and maybe "version" is not the best name. But that's the idea, a kind of application-level once-and-done opt-in for "yes we're ready for these metrics to start flowing now" rather than a runtime user-level opt-in-or-out scenario.
> 
> > Metrics versioning adds another layer of confusion to the analysis, which I'd strongly
> > discourage doing. The tables that will contain the ingested data will contain all of
> > these fields, no matter the version.
> 
> Does the usage described above change the possibility for confusion here? In my head it doesn't seem any different to adding new metrics directly into a new version of an app, but I'm sure there's plenty of nuance involved that I don't really understand well.

I'm afraid not :( This has to do with the way Glean "automagically" builds tables containing your data. The `probe-scraper` fetches the `metrics.yaml` for all the dependencies of Mozilla apps, then assembles a schema for storing data in. It nows about dependencies [from this file](https://github.com/mozilla/probe-scraper/blob/master/repositories.yaml) (which reminds me we need to add logins there, before you can see any data :) ).

> > I would encourage all of us to properly discuss this and make it a Q1 objective.
> 
> 👍 certainly happy to meet and discuss. I don't have a good sense of your team's schedule, would you be happy to set something up with the right people and timezones @Dexterp37? Ideally we can involve @grigoryk but I also feel like I have a good understanding of his requirements and could represent them if necessary.

This feedback is important to us, we'll make time for it. On our end, we can at least commit to discussing and designing a solution. I think @grigoryk and the representative for the product who asked for this requirement should both attend :)


> Which I guess in turn, is what leads to the "The tables that will contain the ingested data will contain all of these fields" situation.

Practically speaking, since the table would have all columns, how would you tell if the data is missing because no collection took place vs the data is missing because the component disabled telemetry? It's complicated, unless we report the active status of telemetry for that component. That would not solve the problem, it would only push this problem to analysts who would need to know one more wrinkle to work with your data :)

Reply:
> They are required to do the data review for adding the Glean component

If the app is already using Glean, and so has done this glean-integration data review, do we intend for it to have to do *another* application-level review for the new metrics from the component?

As currently worded, the [README warnings like this one](https://github.com/mozilla-mobile/android-components/tree/master/components/support/sync-telemetry#before-using-this-component) seem to imply that yes, the application needs an additional review for the new metrics when it starts using that component.

But from conversations around this and related bugs, I think we would like for the answer to be no, and for the component-level data review to be sufficient for inclusion in any Mozilla product past, present or future.

Which I think is what you're getting at here:

> > * If we change the way metrics works in the logins component in the future, we go through the above cycle again
>
> I think that should not be required, unless you merge some concerning metric in your
> component. And, if you do that, we have much bigger problems :-)
> In order to transition this from "I think" to "that's our policy" we reached out to Data
> Stewardship to hopefully encode these practices.

In which case, +1 from me on the developer-experience aspects of this; the policy aspects I will gladly leave to the experts 😁.

> Practically speaking, since the table would have all columns, how would you tell if the
> data is missing because no collection took place vs the data is missing because the
> component disabled telemetry?

I don't really understand how this differs from the case where a new version of an application ships a new metric defined in its top-level `metrics.yaml`. Wouldn't this also cause new columns to appear in the metrics tables for the app, and you have to know which version of the app introduced that metric in order to know how to interpret missing data?

Reply:
Semi-answering my own questions while reading through glean-related docs...

> I don't really understand how this differs from the case where a new version of an
> application ships a new metric defined in its top-level metrics.yaml. Wouldn't this also cause
> new columns to appear in the metrics tables for the app, and you have to know which
> version of the app introduced that metric in order to know how to interpret missing data?

One difference is that you can't use the metadata gathered by probe-scraper in *find out* which version of the app introduced that metric.

Reply:
> 
> 
> > They are required to do the data review for adding the Glean component
> 
> If the app is already using Glean, and so has done this glean-integration data review, do we intend for it to have to do _another_ application-level review for the new metrics from the component?
> 
> As currently worded, the [README warnings like this one](https://github.com/mozilla-mobile/android-components/tree/master/components/support/sync-telemetry#before-using-this-component) seem to imply that yes, the application needs an additional review for the new metrics when it starts using that component.

When the product starts using the Glean-enabled component for the first time (i.e. when it gets integrated), a data-review request must be performed.
Just once.

> But from conversations around this and related bugs, I think we would like for the answer to be no, and for the component-level data review to be sufficient for inclusion in any Mozilla product past, present or future.

Bringing in @chutten, who is a Data Steward who knows both about Glean and, well, data stewardship :)

> Semi-answering my own questions while reading through glean-related docs...
> 
> > I don't really understand how this differs from the case where a new version of an
> > application ships a new metric defined in its top-level metrics.yaml. Wouldn't this also cause
> > new columns to appear in the metrics tables for the app, and you have to know which
> > version of the app introduced that metric in order to know how to interpret missing data?
> 
> One difference is that you can't use the metadata gathered by probe-scraper in _find out_ which version of the app introduced that metric.

Exactly, and that's a deal-breaker. Especially when we need to do things on the pipeline side as changing schemas or backfilling.
This is one thing that you really don't want to break :)

Moreover, that would make it very hard to document which metrics are available in products, which is a requirement. Glean currently does that automatically, with mininal intervention by the product owners to link the docs from the various components.

By going the non-glean way (versioning your metrics), you're required to keep your documentation always up to date, in a format that's human readable (vetted by legal).

This has proven to be a non-trivial task in the past.

Reply:
I wrote in https://bugzilla.mozilla.org/show_bug.cgi?id=1600399#c6, but I'm happy to summarize here: when you do Data Review in your component, you are doing it for the current population your component reaches. In this way the Data Review is much like it is in Firefox Desktop where you assume that, since (say) devtools is included in Firefox that a collection in devtools reaches the population of Firefox.

If someone comes along and integrates devtools into another Application that supplies a data collection library, then that Application is expanding the data collections in devtools to a new population. At that time that new someone performs a Data Review to ensure we document the expansion of collection. From then on, devtools Data Reviews should include in their population response the populations of Firefox Desktop and this new Application.

There is a hole in this where devtools might not know it's being included in Application, so when they do their Data Review they'll miss that they're collecting data from a broader audience than they expect. This is fine. Data Review is forgiving: we put in the best information we have, and if we need to update it later, then we do that. 

And honestly, this is the sort of thing Data Stewards are supposed to help with. Since we did the data review for that Application, you should rely on _us_ to make sure things are right. It's what we do : )

As such, I don't think any component-specific data collection control is needed in specific or in general. From your example, Lockwise upgrading to a new Sync version is fine from Data Stewardship's POV because the new Sync version's data collections will have been reviewed for collection from the Lockwise population. (And if they haven't, then that's Data Stewardship will help amend that. No biggie, just let us know if you notice it and we'll file some bugs to fix it up.)

Does this help @rfk ? @grigoryk ?

Reply:
It does, thank you!

Reply:
@rfk , given @chutten comment, does this PR still make sense?

Reply:
It does not; I'm going to go head and close it down. Thanks @Dexterp37 @chutten!

Reply:
@chutten is it fair to think of this as a trade-off between simplicity and accuracy of the data review process?

Let's take this example, for instance:
> From your example, Lockwise upgrading to a new Sync version is fine from Data Stewardship's POV because the new Sync version's data collections will have been reviewed for collection from the Lockwise population. (And if they haven't, then that's Data Stewardship will help amend that. No biggie, just let us know if you notice it and we'll file some bugs to fix it up.)

It's not clear to me how this is supposed to actually work in practice. Components can and will expand their telemetry collection. Products will upgrade to newer versions of components, or include new components they haven't used before, possibly without filling out data review forms (since developers may not notice a changelog notice, or even look for them, etc). All of this is likely to create a situation in which we're collecting more telemetry than we intend to, for a larger audience than we intend to, without actually realizing that this is happening, or a process in place to reliably prevent this from happening.

I understand that for _updates_ to metrics, from a developer's point of view, a separate review process for each update could be quite onerous, if this happens a lot. It will also place a higher burden on data stewards, as well. But, that seems like a fairly unique case; it's hard to imagine that most components will have churny metrics.

I think one way to phrase my problem with the current situation is that we're over-relying on humans and process, and under-relying on tooling/API design. It's currently completely possible to land telemetry in a new product, that's completely out-of-context for that product, without knowing this happened. Is my assumption true here, or am I missing something that would prevent this from happening?

Another way to look at this, just because telemetry XY is fine to collect within Product A, it's not necessarily the case that it's fine to collect for Product B. Privacy implications/promises of these two products may be quite different!

(slightly offtopic, but this also ties into the larger strategy around components - they exist, in part, to allow us to ship and maintain a variety of different products, supposedly a larger number than we currently have. This hypothetical future of many products, many devs, etc, amplifies the problem.)

Reply:
> I think one way to phrase my problem with the current situation is that we're
> over-relying on humans and process, and under-relying on tooling/API design.

The impression I've come away with is basically: we can and should improve the tooling for this over time (once, in Glean) but the current process won't fail in sufficiently catastrophic ways that we need to block on improved tooling in the meantime.